### PR TITLE
Tests for Norwegian translation

### PR DESCRIPTION
**Pull request by bertfrees**
Includes:
- [x] uncontracted braille
- [ ] 8-dot braille
- [x] contracted braille
- [ ] emphasis indicators

Addresses requirements:
- [x] [4.3:34](http://dev.pef-format.org/dp2/index.xhtml#4.3:34)
- [ ] [4.3:35](http://dev.pef-format.org/dp2/index.xhtml#4.3:35)
- [x] [4.3:37](http://dev.pef-format.org/dp2/index.xhtml#4.3:37)
- [x] [4.3:38](http://dev.pef-format.org/dp2/index.xhtml#4.3:38)
- [ ] [4.3:39](http://dev.pef-format.org/dp2/index.xhtml#4.3:39)

### CODE REVIEW

**Comment by bertfrees** [comment]
For now I've used the provided XML and PEF files for testing emphasis indicators but I see a possible problem with this kind of test data.

The test case contains not only text with emphasized passages, but also various other features such as text indentation, borders, page numbers. This means we have to do more fidgeting to make the test succeed (see for example the style.css file) and more features of course also make the test brittle, so that it's more likely that we'll have to modify the test data or the test later, which means more work. In general, it's better to test more specific features. For example, just a list of input XML snippets vs. expected braille like in http://dev.pef-format.org/dp2/test/43.1/test.xspec.xhtml#43.1 (no PEFs) would already suffice.

Is more test data coming? Currently we have only one example of strong and one example of em. If you think these examples are representative for all your books, that's okay, otherwise I would like to see some more examples.

**Comment by josteinaj** [comment]
I agree, the test data should be as granular as possible.

I think we do have a lot of confidence in liblouis though so if you're ok with this for now then we're ok with this for now.

We haven't got any more data planned at the moment, but I'll look through some old braille documentation and see if there's anything that can be used as test data there.

Maybe it will be easier to create tests after we try out the first test version of the script.

**Comment by bertfrees** [comment]
Sure, creating tests only when something is wrong with the translation is another approach and I'm okay with that. But I also see these tests as some sort of guarantee that things won't break in the future. Especially for liblouis I think we have to keep in mind things *can* break because we don't have full control of it but still we want to get table enhancements done by the community.

File: src/test/38_nlb/test.xprocspec.xsl [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+					<a class="pef-ascii" href="#">ascii</a>
+				</xsl:if>
+			</div>
+		</span>
+	</xsl:template>
+	
+	<xsl:template match="pef:page">
+		<xsl:variable name="rows" select="xs:integer(number(ancestor::*[@rows][1]/@rows))"/>
+		<xsl:variable name="cols" select="xs:integer(number(ancestor::*[@cols][1]/@cols))"/>
+		<!--
```
Comment (bertfrees) [inline src/test/38_nlb/test.xprocspec.xsl]: By the way which Unicode braille <-> ASCII mapping should I use to display Norwegian braille? I believe NLB also uses the Easy Embosser Utility? There should be a similar option in the user interface of EEU.

Comment (josteinaj) [inline src/test/38_nlb/test.xprocspec.xsl]: I don't know. I'll check with Kari. What are the alternatives?

Comment (bertfrees) [inline src/test/38_nlb/test.xprocspec.xsl]: I think it should be one of these: https://github.com/mtmse/brailleutils/blob/master/braille-utils.impl/src/org_daisy/EmbosserTableProvider.java. But I couldn't find Norwegian in there. We can also create a new mapping of course, and I could even use the liblouis display table [no-no.dis](https://github.com/liblouis/liblouis/blob/master/tables/no-no.dis).

Comment (josteinaj) [inline src/test/38_nlb/test.xprocspec.xsl]: According to @indy78 we use Unicode Braille.

Comment (bertfrees) [inline src/test/38_nlb/test.xprocspec.xsl]: Okay, thanks. But can you really read that??

Comment (josteinaj) [inline src/test/38_nlb/test.xprocspec.xsl]: I have no idea. Is this easy to change in the future?

Comment (bertfrees) [inline src/test/38_nlb/test.xprocspec.xsl]: Of course, whenever you want.

Comment (josteinaj) [inline src/test/38_nlb/test.xprocspec.xsl]: Cool :)

**Comment by bertfrees** [comment]
I referenced an external liblouis test (see [4.3:34](http://dev.pef-format.org/dp2/norwegian/index.xhtml#4.3:34)) and added a test for text-level semantics (see [4.3:38](http://dev.pef-format.org/dp2/norwegian/index.xhtml#4.3:38)).

**Comment by bertfrees** [comment]
There is currently no definition of 8-dot braille. Need to find out whether defining and/or implementing are in scope. If yes, what is the priority.

**Comment by bertfrees** [comment]
New test data for uncontracted and contracted are available at [no_harness.txt](https://github.com/snaekobbi/liblouis/blob/7e3dc262f71b6a6bc050620cf170943807977616/tests/harness/no_harness.txt), [no_harness_g1.txt](https://github.com/snaekobbi/liblouis/blob/7e3dc262f71b6a6bc050620cf170943807977616/tests/harness/no_harness_g1.txt) and [no_harness_g1.txt](https://github.com/snaekobbi/liblouis/blob/7e3dc262f71b6a6bc050620cf170943807977616/tests/harness/no_harness_g2.txt)

**Comment by bertfrees** [comment]
@usama49 While you are looking at https://github.com/snaekobbi/pipeline-mod-braille/issues/41 you can maybe also have a look at requirement [4.3:39](http://dev.pef-format.org/dp2/index.xhtml#4.3:39) and check if any tests need to be added for this requirement (could be added to [test_translator.xprocspec](https://github.com/snaekobbi/pipeline-mod-nlb/blob/cb2ce6230719e8725c15fbd23b59231dfb793228/src/test/xprocspec/test_translator.xprocspec))

**Pull request merged**